lib/rucio/rse/protocols/rclone.py

```diff
@@ -0,0 +1,388 @@
+# -*- coding: utf-8 -*-
+# Copyright 2013-2020 CERN
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Rakshita Varadarajan <rakshitajps@gmail.com>, 2021
+
+import os
+import logging
+import json
+
+from rucio.common import exception
+from rucio.rse.protocols import protocol
+from rucio.common.utils import execute, PREFERRED_CHECKSUM
+from rucio.common.config import get_config_dirs, get_rse_credentials
+
+
+def load_conf_file(file_name):
+    config_dir = next(filter(lambda d: os.path.exists(os.path.join(d, file_name)), get_config_dirs()))
+    with open(os.path.join(config_dir, file_name)) as f:
+        return json.load(f)
+
+
+class Default(protocol.RSEProtocol):
+    """ Implementing access to RSEs using the rclone protocol."""
+
+    def __init__(self, protocol_attr, rse_settings, logger=logging.log):
+        """ Initializes the object with information about the referred RSE.
+
+            :param props Properties derived from the RSE Repository
+        """
+        super(Default, self).__init__(protocol_attr, rse_settings, logger=logger)
+        if len(rse_settings['protocols']) == 1:
+            raise exception.RucioException('rclone requires atleast one other base protocol to be initialized upon.')
+        self.scheme = self.attributes['scheme']
+        setuprclone = False
+        for protocols in reversed(rse_settings['protocols']):
+            if protocol_attr['impl'] == protocols['impl']:
+                continue
+            else:
+                setuprclone = self.setuphostname(protocols)
+                if setuprclone:
+                    break
+
+        if not setuprclone:
+            raise exception.RucioException('rclone could not be initialized due to insufficient required data.')
+        self.logger = logger
+
+    def setuphostname(self, protocols):
+        """ Initializes the rclone object with information about protocols in the referred RSE.
+
+            :param protocols Protocols in the RSE
+        """
+        if protocols['scheme'] in ['scp', 'rsync', 'sftp']:
+            self.hostname = 'ssh_rclone_rse'
+            self.host = protocols['hostname']
+            self.port = str(protocols['port'])
+            if protocols['extended_attributes'] is not None and 'user' in list(protocols['extended_attributes'].keys()):
+                self.user = protocols['extended_attributes']['user']
+            else:
+                self.user = 'root'
```

Comment:
:thinking:

Reply:
@mlassnig default value that i had set for the user accessing the SSH server, should i instead set it to "", and if user value is present then assign self.user that value?